### ISSUE TITLE
Inlined nonterminals

### ISSUE THREAD

**Issue by nikomatsakis** [nikomatsakis/lalrpop#15]
### Checklist

- [x] Add annotations, make it a validation error to use anything but `#[inline]` 
- [ ] Do ininling expansion, either at the end of normalization or (perhaps better) on the `repr` itself

### Description

One of the features I've wanted to have for a long time is inlined nonterminals. The idea is that if you declare a terminal as `#[inline]`:

 ```rust
#[inline]
A = B | C | ();
```

then every use of `A` will be expanded out to its various alternatives. In other words, a reference like:

```rust
X = A "foo";
```

would become:

```rust
X = B "foo" | C "foo" | "foo";
```

It has been my experience that inlining nonterminals is very helpful for removing shift-reduce conflicts, particularly around optional content. Accordingly, the `X*` and `X?` shorthands would expand to inlined nonterminals.

This should solve shift-reduce conflicts like:

```rust
T = "L" | "&" "L"? T
```

Here we get a conflict in the state `& (*) L`, where it's unclear whether the `L` is the optional `L` or part of the `T`. In the latter case, we'd have to reduce `() => "L"?`. But if we inlined the `"L"?` expansion, we'd have:

```rust
T = "L" | "&" T | "&" "L" T
```

In which case, we could just shift the `"L"` and wait until see the next token to decide what to do.

User annotation is needed here because inlining affects the visible order of execution.

Note though that inlining is only really useful in an `LR(1)` setting. If we chose to adopt some kind of universal algorithm by default this would become less important.

**Comment by nikomatsakis** [comment]
I've got this basically implemented in the `inline` branch

**Comment by nikomatsakis** [comment]
Fixed by #36
